### PR TITLE
GH-44606: [C++] Add find module for abseil that handles missing version

### PR DESCRIPTION
**Pull request by assignUser**
### Rationale for this change
Abseil's recommended usage is to build from trunk, this version does not provide a package version in CMake which breaks `resolve_dependency`.

### What changes are included in this PR?

Add `FindabslAlt.cmake` that adds a big version if it is missing to compare against our minimal version requirements.

### Are these changes tested?

CI

### Are there any user-facing changes?

Abseil build from head should now work without workaround.
* GitHub Issue: #44606

### CODE REVIEW

**Comment by assignUser** [comment]
@github-actions crossbow submit -g cpp r-binary-packages

**Comment by assignUser** [comment]
I don't think another crossbow run will give us anything more than  the normal PR CI.

File: cpp/cmake_modules/FindabslAlt.cmake [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+
+set(find_package_args)
+
+if(abslAlt_FIND_VERSION)
+  list(APPEND find_package_args ${abslAlt_FIND_VERSION})
+endif()
+if(abslAlt_FIND_QUIETLY)
+  list(APPEND find_package_args QUIET)
+endif()
+
```
Comment (kou) [inline cpp/cmake_modules/FindabslAlt.cmake:30]:
```suggestion
```

File: cpp/cmake_modules/FindabslAlt.cmake [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+if(abslAlt_FOUND)
+  return()
+endif()
+
+set(find_package_args)
+
+if(abslAlt_FIND_VERSION)
+  list(APPEND find_package_args ${abslAlt_FIND_VERSION})
+endif()
```
Comment (kou) [inline cpp/cmake_modules/FindabslAlt.cmake:26]: Do we need this?

File: cpp/cmake_modules/FindabslAlt.cmake [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+if(abslAlt_FIND_REQUIRED)
+  list(APPEND find_package_args REQUIRED)
+endif()
+
+find_package(absl ${find_package_args})
+
+if(NOT DEFINED absl_VERSION)
+  # Abseil does not define a version when build 'live at head'.
+  # As this is their recommended path we need to define a large version to pass version checks.
+  set(absl_VERSION 99999999_head)
```
Comment (kou) [inline cpp/cmake_modules/FindabslAlt.cmake:40]: Do we need the `_head` part?

```suggestion
  set(absl_VERSION 99999999)
```

**Review by kou — APPROVED** [review]
+1

File: cpp/cmake_modules/FindabslAlt.cmake (hunk 3 above)
Comment (assignUser) [reply cpp/cmake_modules/FindabslAlt.cmake:40]: Technically not but the absl.pc uses 'head' as a version and it doesn't influence the comparision with our minum version and is shown in the configure log.

File: cpp/cmake_modules/FindabslAlt.cmake (hunk 2 above)
Comment (assignUser) [reply cpp/cmake_modules/FindabslAlt.cmake:26]: As we use it currently not, I guess it could produce confusing results, let's remove it.

Comment (assignUser) [inline cpp/cmake_modules/FindabslAlt.cmake:26]:
```suggestion
```

File: cpp/cmake_modules/FindabslAlt.cmake (hunk 3 above)
Comment (kou) [reply cpp/cmake_modules/FindabslAlt.cmake:36]: OK.

It seems that the `VERSION_*` operators remove the `_head` part:

https://cmake.org/cmake/help/latest/command/if.html#version-comparisons

> (version format is major[.minor[.patch[.tweak]]], omitted components are treated as zero). Any non-integer version component or non-integer trailing part of a version component effectively truncates the string at that point.

Could you add a comment that the `_head` part is safe for version comparison?

**Review by kou — APPROVED** [review]
+1

**Pull request merged**